### PR TITLE
Create pharmacy migrate split and migrations.

### LINKED ISSUE #2987: Pharmacy news migration
# Story

As a College of Pharmacy website owner, I would like news articles migrated from our existing website (https://pharmacy.uiowa.edu") to our new website (https://pharmacy.prod.drupal.uiowa.edu), so that site visitors can see past Pharmacy related news without the College of Pharmacy having to manually create hundreds of old news articles manually.

# Additional Info

>We would like to migrate all historical news articles from our current website to our new website, while also preserving backlinks. We have articles dating back to 2012.  See: https://pharmacy.uiowa.edu/news/archive

We need to create redirects from old paths to new as part of this migration.
 
>Ideally, we would also like hyperlinks removed from articles from 2016 and earlier. That is, links that appear within the articles themselves.

They are fine with just removing the links as part of the migration process, no need to provide a list of node id's that we removed links from.

# Proposed solution

### LINKED ISSUE #2999: As a visitor to pharmacy.uiowa.edu, I can get redirected from old directory person pages to new directory person pages.
# Story

As a visitor to pharmacy.uiowa.edu, I can get redirected from old directory person pages to new directory person pages.

# Problem
The current pharmacy directory nests person pages under /directory/person/{slug} while the new APR directory only nests them under whatever is configured as the the directory path. Webmasters could create redirects for each person but that would tedious. We could create the redirect in the htaccess file but if the destination ever changes, the redirects will break.

# Proposed solution

Investigate ways for webmasters to bulk import lots of redirects or create wildcard redirects that forward the path along to the destination. If thats not feasible, create the redirect in htaccess with the caveat that the APR directory config cannot change without notifying us.

### LINKED ISSUE #3182: Enable remote post handler for Pharmacy webforms
# Story

As a College of Pharmacy website owner, I want the ability to post my webform submissions to my CRM, so that we don't have to download and/or re-enter the data manually.

# Proposed solution

### PR DESCRIPTION
**Pull request by joewhitsitt**
This PR has been hijacked. It was Initially for #3182. 

Resolves #2987
Resolves #2999

### Testing
TLDR; https://pharmacy.dev.drupal.uiowa.edu

- Sync D7 pharmacy site in DD.
```
blt sync --site pharmacy.uiowa.edu && 
drush @pharmacy.local en pharmacy_migrate --yes && 
drush @pharmacy.local uli /admin/config/system/sitenow-migrate &&
drush @pharmacy.local uli admin/config/system/sitenow-migrate
```
- Configure sitenow_migrate settings to connect to D7 site.
- `drush @pharmacy.local cr`
- `drush @pharmacy.local migrate:status`
- `drush @pharmacy.local migrate:import --all`
- If the migration is marked as a failure with `Unkown format` errors, run `drush @pharmacy.local migrate:import --all` again and it should finish.
- Checks
  - Should have aligned-left, inline images using small__no_crop: https://pharmacy.dev.drupal.uiowa.edu/news/2018/09/distinguished-alumni-award-recipients-announced
  - Should have no links: http://pharmacy.local.drupal.uiowa.edu/news/2015/02/pharmacy-leaders-explore-future-pharmacy-technician-education-iowa
  - Should have some person links pointing to new location: http://pharmacy.local.drupal.uiowa.edu/news/2020/11/second-year-flip-pharmacy-include-covid-19-vaccines Ex. William Doucette

### CODE REVIEW

**Comment by richardbporter** [comment]
Do they want to test this in dev first? I'm not sure if thats possible. Do you think we'd have to push changes for this to get it all working?

**Comment by briand44** [comment]
They haven't launched their site yet so if this allows them to add a remote post handler then I think we can just deploy it to prod and let them try it out prior to them launching. Since they want to launch sometime in April, I am hoping they can handle the post data on their CRM side of things. If it ends up requiring further support on our end to get this working for them, that would likely have to come sometime after launch and they could continue using HubSpot until we have capacity to work on this further.

**Comment by richardbporter** [comment]
Tested with a custom post bin and seems to work. This is slightly concerning though as it shows dev info for all users including anonymous:

image

Seems like if you are posting to a custom endpoint, debugging on the Drupal side wouldn't be that useful. Might expose something we don't want to.

**Comment by richardbporter** [comment]
Can we hide that? I think we're doing a bit of that in sitenow.profile or somewhere already.

**Comment by richardbporter** [comment]
Some of the more advanced options might let webmasters mess things up but we've already ran into that with current YAML options and such. Nothing new there.

**Comment by briand44** [comment]
If there are security concerns with the debugging output, I think it is fine to hide this option for all roles other than administrator.

**Comment by joewhitsitt** [comment]
>I think it is fine to hide this option for all roles other than administrator.

Not sure I saw that permission

**Comment by briand44** [comment]
I guess I wasn't assuming there was a permission for that. Assumed we would do it in code based on this comment.
>Can we hide that? I think we're doing a bit of that in sitenow.profile or somewhere already.

**Comment by joewhitsitt** [comment]
Ah. Missed that comment. I can look into this when I can.

**Comment by briand44** [comment]
Sorry, I just got word from the customer that it was determined internally that they want to continue using the HubSpot forms for now and just embed them on their new site. No further work is necessary on this at this time. We can probably close this issue and reopen it if/when they decide to pursue this further.

**Comment by joewhitsitt** [comment]
Keeping this around as I imagine pharmacy will go custom eventually with or without this remote post functionality.

**Comment by richardbporter** [comment]
Could we still close the PR and reopen it later whenever that happens?

**Comment by joewhitsitt** [comment]
Closed PRs' branches tend to get deleted but I understand the desire to clean up the PR list. Fine with whatever.

**Comment by richardbporter** [comment]
I think closed PR branches can be restored - not seeing a time limit in GitHub docs - so I assume indefintely.

**Comment by richardbporter** [comment]
Keep seeing this error importing images and I've no clue how to debug it: ` [error]  Unknown format: 0x300 `

**Comment by richardbporter** [comment]
The engineering and grad migrations will need to be updated to match the sitenow_migrate changes if they need to run again.

**Comment by richardbporter** [comment]
I think this ready for code review/testing but we probably need to wait for pharmacy sign off. I ran the migrations in dev and pinged them.

**Comment by joewhitsitt** [comment]
pulling down to test

**Comment by joewhitsitt** [comment]
Roughing it with these rough test instructions, but:

```
drush @pharmacy.local migrate:import --all
 [notice] Processed 988 items (988 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_people_redirects'
 [notice] Processed 12 items (12 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_tags'
 159/650 [======>---------------------]  24% [error]  Unknown format: 0x300 
 [error]  Unknown format: 0x300 
 184/650 [=======>--------------------]  28% [error]  Unknown format: 0x300 
 458/650 [===================>--------]  70% [error]  Unknown format: 0x300 
 [error]  Unknown format: 0x300 
 474/650 [====================>-------]  72% [error]  Unknown format: 0x300 
 476/650 [====================>-------]  73% [error]  Unknown format: 0x300 
 495/650 [=====================>------]  76% [error]  Unknown format: 0x300 
 498/650 [=====================>------]  76% [error]  Unknown format: 0x300 
 499/650 [=====================>------]  76% [error]  Unknown format: 0x300 
 517/650 [======================>-----]  79% [error]  Unknown format: 0x300 
 535/650 [=======================>----]  82% [error]  Unknown format: 0x300 
 541/650 [=======================>----]  83% [error]  Unknown format: 0x300 
 569/650 [========================>---]  87% [warning] Memory usage is 436.33 MB (85% of limit 512 MB), reclaiming memory.
 [warning] Memory usage is now 434.03 MB (85% of limit 512 MB), not enough reclaimed, starting new batch
 [notice] Processed 569 items (569 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_articles'
 [error]  Migration pharmacy_article_redirects did not meet the requirements. Missing migrations pharmacy_articles. requirements: pharmacy_articles. 

In MigrateToolsCommands.php line 866:
                                                
  pharmacy_article_redirects migration failed. 
```

Spot checked the person redirect. Seemed to work but my APR people aren't showing locally. I assume that is a local issue.

News seems to be imported for the most part. Broken images though. Seems like some news articles just contain a link to a node or the date isn't right? http://pharmacy.local.drupal.uiowa.edu/news/2020/03/ernsts-stewardship-combats-superbugs https://pharmacy.uiowa.edu/article/ernsts-stewardship-combats-superbugs

Screen Shot 2021-03-30 at 11 03 46 AM

**Comment by richardbporter** [comment]
Stuck in #3288 if someone else wants to investigate these issues.

**Comment by richardbporter** [comment]
I thought the `[error]  Unknown format: 0x300 ` messages were memory-related and I implemented `clearMemory` after some refactoring which cleared it up for me. https://github.com/uiowa/uiowa/pull/3190/files#diff-ed54d7d976930eb74a1a63ef0ff7bb39f2f5331459760eb989902c9dd5a02ff8R45 Could be wrong though as that message is pretty cryptic. Maybe @cory-skeers has seen it before?

I'm looking into the other issues.

**Comment by richardbporter** [comment]
> News seems to be imported for the most part. Broken images though. Seems like some news articles just contain a link to a node or the date isn't right? http://pharmacy.local.drupal.uiowa.edu/news/2020/03/ernsts-stewardship-combats-superbugs https://pharmacy.uiowa.edu/article/ernsts-stewardship-combats-superbugs

That article looks to have been manually created on the new site and was not one that imported. Not sure why they did that. The actual imported article is http://pharmacy.local.drupal.uiowa.edu/news/2016/07/ernsts-stewardship-combats-superbugs.

@joewhitsitt let me know if there are other examples, especially for the broken images.

**Comment by joewhitsitt** [comment]
i will take another look

**Comment by joewhitsitt** [comment]
Might be a local thing.

```
583/650 [=========================>--]  89% [warning] Memory usage is 435.32 MB (85% of limit 512 MB), reclaiming memory.
 [warning] Memory usage is now 434.43 MB (85% of limit 512 MB), not enough reclaimed, starting new batch
 [notice] Processed 583 items (583 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_articles'
 [error]  Migration pharmacy_article_redirects did not meet the requirements. Missing migrations pharmacy_articles. requirements: pharmacy_articles. 

In MigrateToolsCommands.php line 866:
                                                
  pharmacy_article_redirects migration failed.  
```

I must have picked some odd ones to spot check. I do see a lot of broken images. My files/2021-03 folder only has 292 items.

Screen Shot 2021-03-31 at 1 44 59 PM

**Addendum:** Looks like DEV is good. If Pharmacy's QA check is "all accounted for" then I am good.

**Comment by joewhitsitt** [comment]
Sean pointed out it might be because of stage_file_proxy not being wired up. I assumed the migration would be pulling article related files locally to my machine

**Comment by richardbporter** [comment]
I'll check it out again.

**Comment by richardbporter** [comment]
I can't get past those memory errors, assuming thats even what it is. I had to comment out both processMediaTrait functions to get them to stop so its narrowed down a bit.

I think I had to run this in very small increments on dev over and over and over again to get it to run.

 Still digging...

**Comment by cory-skeers** [comment]
> Sean pointed out it might be because of stage_file_proxy not being wired up. I assumed the migration would be pulling article related files locally to my machine

Might have changed, but the media processing should check if there's already a file/media with a given filename listed in the database, and use that if available before downloading. So if there are images already on the prod site that are used in articles, those will need stage file proxy to work.

**Comment by pyrello** [comment]
It seems like some of the memory clearing techniques we are using should already be in place in core... https://www.drupal.org/project/drupal/issues/2503999

**Comment by cory-skeers** [comment]
> It seems like some of the memory clearing techniques we are using should already be in place in core... https://www.drupal.org/project/drupal/issues/2503999

Yep--it's the same as what core does to (attempt to) clear memory when it approaches running out. With the engineering migrate there were points when it would go from not using enough memory to trigger core cleanup to running out of memory before the next check. The `clearMemory` in `BaseNodeSource` is just any easy way to call it on demand.

> I can't get past those memory errors, assuming thats even what it is. I had to comment out both processMediaTrait functions to get them to stop so its narrowed down a bit.

> I think I had to run this in very small increments on dev over and over and over again to get it to run.

That's about as far as I got, as well as some testing on handling things different ways, clearing memory within the function, and minor things like NULLing file data variables to get them out of memory right away.

The problem I ran into with engineering taxonomy terms being duplicated was a result of needing to incrementally run the migration--had been dynamically creating new taxonomy terms and holding them in an array, and didn't account for the fact re-running it would be starting over each time... As long as there's no per-row processing like that going on, there shouldn't be any issues with running it incrementally other than time and annoyance.

**Comment by pyrello** [comment]
I'm not sure if this is how we have done it for other migrations since I was last working on one, but I'm not sure about the approach of having the migration split config in the site split itself for pharmacy. It seems like this forces us to a.) Later remove those elements from the split to turn off the migration modules, or b.) Leave the migration modules turned on.

In previous cases I had created a separate split for the migration that could be turned on and off, thus uninstalling the relevant modules. This also make eventual removal easier because we don't have to update and re-export the site split, we can just remove the site's migration split.

**Comment by pyrello** [comment]
One thing I'm noticing that is weird: When I copy the address for the image that is not showing up and paste it into the browser, I get redirected to the current pharmacy site.
http://pharmacy.local.drupal.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV
=>
https://pharmacy.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV

**Comment by pyrello** [comment]
I was able to run the migration without any issues or errors:
```
 [notice] Processed 984 items (984 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_people_redirects'
 [notice] Processed 12 items (12 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_tags'
 [notice] Processed 650 items (650 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_articles'
 [notice] Processed 650 items (650 created, 0 updated, 0 failed, 0 ignored) - done with 'pharmacy_article_redirects'
```

Since I didn't encounter any memory problems, I'm wondering if this is an environmental issue? Why would memory problems crop up some of the time and not others?

Also, I didn't hit any of the image format errors. This makes me wonder if, when other people have been testing this, have you cleared out your files prior to re-running it? I'm not sure that this causes any issues, but it can definitely affect the results, so I got in the habit of removing files prior to running the migrations.

**Comment by cory-skeers** [comment]
> One thing I'm noticing that is weird: When I copy the address for the image that is not showing up and paste it into the browser, I get redirected to the current pharmacy site.
> http://pharmacy.local.drupal.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV
> =>
> https://pharmacy.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV

I'm not super clear on how stage file proxy is setup, but `lockman.jpg` is an image that should already exist on the site (https://pharmacy.prod.drupal.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV). Whether the redirect is intended behavior or not, don't know, but might narrow down the weirdness.

**Comment by richardbporter** [comment]
> Later remove those elements from the split to turn off the migration module

I don't see that being much more work but don't really care which we do it.

**Comment by pyrello** [comment]
> I'm not super clear on how stage file proxy is setup, but lockman.jpg is an image that should already exist on the site (https://pharmacy.prod.drupal.uiowa.edu/sites/pharmacy.uiowa.edu/files/styles/widescreen__768_x_432/public/2021-03/lockman.JPG?h=3e51a2f9&itok=AufvcwEV). Whether the redirect is intended behavior or not, don't know, but might narrow down the weirdness.

D'oh! I had SFP configured to point to pharmacy.uiowa.edu instead of pharmacy.prod.drupal.uiowa.edu. After I changed that, it looks okay. I'm not seeing any issues at all.

**Comment by pyrello** [comment]
> I'm not seeing any issues at all.

Scratch that. I'm seeing some HTML getting rendered as text:

Screen Shot 2021-04-01 at 1 41 49 PM
Screen Shot 2021-04-01 at 1 42 04 PM

**Comment by cory-skeers** [comment]
> Scratch that. I'm seeing some HTML getting rendered as text:

If you edit the page and save, does it fix? If so, it's likely the text field format not getting set or not sticking (saving will set it, even if you don't manually change it).

If not, I can try taking a look at it too.

**Comment by pyrello** [comment]
It seems like this might be related to this section of code:
https://github.com/uiowa/uiowa/blob/3c928f3fdb0d7dd40b48c4dd2146a2e10f4ec6ed/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/Articles.php#L35-L41

Since the [extraSummaryFromText](https://github.com/uiowa/uiowa/blob/3c928f3fdb0d7dd40b48c4dd2146a2e10f4ec6ed/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/BaseNodeSource.php#L100) function seems like it should be stripping these tags, I guess I'm thinking that it is hitting this condition:
https://github.com/uiowa/uiowa/blob/3c928f3fdb0d7dd40b48c4dd2146a2e10f4ec6ed/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/BaseNodeSource.php#L227-L229

This is some of the regex that I don't entirely understand.

**Comment by pyrello** [comment]
@cory-skeers Saving it doesn't fix it because the summary field doesn't use a format:
Screen Shot 2021-04-01 at 1 57 48 PM

**Comment by cory-skeers** [comment]
> @cory-skeers Saving it doesn't fix it because the summary field doesn't use a format:

Ahhhh...Sorry, missed that it was a summary field.

**Comment by pyrello** [comment]
> I don't see that being much more work but don't really care which we do it.

I don't have a strong feeling about this other than that this change in approach makes me nervous.

**Comment by richardbporter** [comment]
Ok, I left #3288 using a feature split so we'll need to update this PR to do something similar. I guess it was kinda pointless for me to resurrect this PR... 🤦🏻

**Comment by pyrello** [comment]
I updated this to:
* Move config for migration to `pharmacy_migrate/config/split`.
* I added a `pharmacy_migrate.install` file to install the config when the module gets turned on. It occurs to me that maybe this config could just go in `config/install`. I don't remember exactly why I set it up this way before.

I also updated the testing instructions.

**Comment by cory-skeers** [comment]
The html appearing in summary was an order of operations issue, looks like. We were trimming the text prior to removing tags, so occasionally we'd cut off the end of a tag, preventing it from being properly stripped. Also led to an issue where if text began with long enough tags, the constructed summary would end up being blank, defeating the purpose of attempting to create one.

**Comment by richardbporter** [comment]
I think I narrowed down the `Unknown format: 0x300` error to the `file_save_data` function call in `ProcessMediaTrait::downloadFile`. Not sure why yet.

**Comment by richardbporter** [comment]
Also -  I wasn't sure where to set that image style for the body images.

**Comment by cory-skeers** [comment]
> I think I narrowed down the `Unknown format: 0x300` error to the `file_save_data` function call in `ProcessMediaTrait::downloadFile`. Not sure why yet.

Yay! I did notice that on rolling back and then rerunning the migration, I didn't run into any of those errors, so that makes sense--it wouldn't need to re-download.

Re: the image style, I thought it would be in https://github.com/uiowa/uiowa/blob/c05630523a8d365227edb66a78d40195da167a2e/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/ProcessMediaTrait.php#L448-L453, but when I tried setting it directly there, it doesn't seem to stick.

**Comment by pyrello** [comment]
> Also - I wasn't sure where to set that image style for the body images.

I'm not sure how we ended up at the point where it is commented out, but this appears to be where it would have been set:
https://github.com/uiowa/uiowa/blob/c05630523a8d365227edb66a78d40195da167a2e/docroot/modules/custom/sitenow_migrate/src/Plugin/migrate/source/ProcessMediaTrait.php#L452

**Comment by richardbporter** [comment]
Source NID 3144 "College of Pharmacy Welcomes Two New Faculty Members" with body image https://pharmacy.uiowa.edu/sites/pharmacy.uiowa.edu/files/wysiwyg_uploads/goedken.jpg throws this error.

**Comment by richardbporter** [comment]
I must have broken something refactoring all this. Seeing a lot of errors after 8d97a44. Leaving it for debug right now.

**Comment by richardbporter** [comment]
I'm assuming the `Unknown format` error is what marks the migration as failed. However, it doesn't seem to leave the migration status as running so it doesn't require resetting it. You can just run `drush @pharmacy.local mim --all` again. I updated the testing instructions.

**Comment by richardbporter** [comment]
I don't have any other immediate work planned for this so marking ready for review again. I re-ran it in dev and pinged the pharmacy team for review there.

**Comment by joewhitsitt** [comment]
giving this another go

**Comment by joewhitsitt** [comment]
I don't know why the partial import step was removed from the test instructions. I need to do that or the migrate status was empty. pharmacy_migrate.install didn't seem to do what it was supposed to do and has "grad" language included.

**Comment by pyrello** [comment]
What I had to do was enter the DB information, then run `drush @pharmacy.local cr`. After that, the migrations show up without the additional config import.

**Comment by pyrello** [comment]
Updated test instructions.

**Comment by cory-skeers** [comment]
Interesting...I don't have to do the extra import, or the cache rebuild.

Fwiw, I fixed the `grad_migrate` language that was left over!

**Comment by joewhitsitt** [comment]
Spot checking and it seems that markup is being added to the plain text summary:

Screen Shot 2021-04-07 at 11 57 15 AM
Screen Shot 2021-04-07 at 11 56 49 AM
Screen Shot 2021-04-07 at 11 56 56 AM

**Comment by cory-skeers** [comment]
I'm not seeing the markup on my run.

image

There was some odd spacing, like in the third example you provided, that should be a quick fix with an `lstrip()`. (Ran into an issue that's not letting me push the change up, but working on it...)

image

**Comment by richardbporter** [comment]
Just heard back. Draft...

**Comment by richardbporter** [comment]
I recommended they hide article images site-wide and enable it for articles where the image looks good. I don't think there is anything to do here.

**Review by richardbporter — APPROVED** [review]
Pharmacy gave this the green light. 👍🏻 

Thanks everyone.
